### ISSUE TITLE
Bug: Linter fails to recognize aliases containing commas

### ISSUE THREAD

**Issue by ScottKillen** [platers/obsidian-linter#525]
## Describe the Bug

This is related to the fix for #509

The following YAML array contains two aliases: 

`aliases: Scott, "Scott, Jr."`

Linter parses this into three aliases.

## How to Reproduce

I have attached the [data.json](https://github.com/platers/obsidian-linter/files/10134799/data.json.txt) file.

See the example above.

Linter produces a multiline array of three aliases:

```yaml
aliases:
  - Scott
  - '"Scott'
  - 'Jr."'
```

## Expected Behavior

Linter should parse the tags array exactly how Obsidian parses the tags array.

The example should produce a multiline tags array with two tags. The double quotes should be eliminated, and the tag containing the commas should not be escaped with single quotes.

Expected output:

``` yaml
aliases:
  - Scott
  - Scott, Jr.
```

## Device

- [x] Desktop
- [?] Mobile

**Comment by ScottKillen** [comment]
I was thinking about this a little more. A good test for Linter would be that properly formed single-line arrays can be converted to multiline arrays and then back to single-line arrays with no change. Likewise, proper multiline arrays should convert to single-line arrays and back to multiline with no change.

This bug would fail these tests, Linter's array conversion should be symmetrical.

**Comment by pjkaufman** [comment]
Thanks for reporting this. This is a known limitation of parsing the values that are not actually a valid format for YAML (I am not sure if it is in the documentation for the rules themselves). This is definitely an issue, but unless the logic for parsing YAML is written to be more robust, this is not necessarily something that can be added.

I am open to testable fixes to this as using the Obsidian built in mechanisms for parsing these values is not really testable and thus is not a reliable solution for us.

**Comment by pjkaufman** [comment]
The main problem with what you are proposing is that the above is not valid YAML so converting from single-line to multi-line is not feasible for it. If it were valid YAML, it would be simple to convert between the two formats. However Obsidian has added for several formats that are not valid, so support for them is kind of monkey-patched in. I definitely am open to improving how things are parsed, but I think it should cover most cases and not all cases as some of them are so infrequent it may not be worth the effort to try to cover.

**Comment by pjkaufman** [comment]
Here are a couple of things to keep in mind for this. 

> Obsidian recognizes the single-line example as valid YAML. As does https://jsonformatter.org/yaml-validator.
>
> Linter's job is to make sure that a file is properly formatted--so if Obsidian recognized two aliases before running Linter, it should recognize only two aliases after Linter is done. 
> Linter is the source of the single-line array example (from my original multiline array) then, when I asked Linter to convert it back to a multi-line array, it could not parse its own emitted YAML properly.
>
> Linter should be able to parse its own results.

While it would be nice if the Linter could parse its own output, the Linter does not posses a means to parse all valid forms of YAML for purposes they were not designed for. 
For example, Obsidian adds at least 3 kinds of new arrays that are not valid YAML arrays, but are still valid YAML strings: 
- comma separated single string
- space separated single string
- single-line array that is space delimited

The only two array formats that YAML recognizes are 
- single-line array
- multiline array

That is what causes some of the problem as YAML parser packages are not designed to parse out the Obsidian based arrays and Obsidian provides something to do the parsing that is not usable for testing which means it can change with no way for devs to know ahead of time. This means that a custom parser has to be used to emulate how Obsidian does its parsing. This is where the more simplistic approach is used. That is what you are seeing with the unexpected output for array differences. It definitely can be improved.

TLDR;
The YAML is being custom parsed which definitely can be improved. Custom parsing is used to recognize the new custom formats for YAML arrays that are not traditionally recognized by YAML parsers as arrays. I am happy to improve the parsing logic, but it will likely never work exactly like Obsidian does in all cases. It will just act like it in most cases.

**Comment by pjkaufman** [comment]
I may have something working for this. It is relatively basic and does not allow nesting arrays, so that would be something to look out for. However it should be able to do a basic parsing of arrays. It is currently allowing all UTs to pass, so I will start adding tests for the cases where there are escaped delimiters and see what happens.

**Comment by pjkaufman** [comment]
Parser seems to be doing its job at determining what should and should not be considered escaped. I have to still add some edge cases, but I am hopeful it will work in those cases. Once that is done, I will go ahead and add a couple UTs for formatting YAML and then push it up for the next release.

Hopefully it all goes well and then I can focus on this header increment stuff.

**Comment by pjkaufman** [comment]
It looks like it plays well with the change. @ScottKillen , could you clarify why you would expect the Linter to automatically remove the double quotes from around `Scott, Jr.` when it goes to multi-line? I know that it no longer needs them, but removing escaping is actually not as straight forward as adding them and I am not sure the Linter removes escaping at this time.

**Comment by pjkaufman** [comment]
As a note, the change for this has been merged into master, so the fix should be present in on master and in the next release. I have not closed this issue as I am curious what the feedback is on removing the escaping of the comma. I am guessing I can add an extra param in somewhere that says that if you are going to become a multi-line array it should remove the escaping of commas, but I want to make sure that that is the rationale here before I move forward with any kind of change like that.

**Comment by ScottKillen** [comment]
> It looks like it plays well with the change. @ScottKillen , could you clarify why you would expect the Linter to automatically remove the double quotes from around `Scott, Jr.` when it goes to multi-line? I know that it no longer needs them, but removing escaping is actually not as straight forward as adding them and I am not sure the Linter removes escaping at this time.

@pjkaufman If the quotes are present everything works fine. The reason I would expect them to be removed is that they are not necessary, which is kind of Linter's thing--and Linter is smart enough to **add** them when they are necessary.

**Comment by pjkaufman** [comment]
Gotcha. I will think on this and see if it makes sense and is feasible to remove the quotes for multi-line and other array formats. It just gets a little hard to tell when quotes should be removed without knowing the previous and new formats of the array.

**Comment by pjkaufman** [comment]
Would it make sense to make this another one of those global settings that would be reusable throughout rules to allow for the removal of the unnecessary quotes for multiline arrays?

I have been debating whether to make this a global setting, automatic, or add it on a rule by rule basis. Since I have not come to a decision, I decided to ask and see what you think @ScottKillen.

**Comment by ScottKillen** [comment]
I think a global setting makes sense. Some may want the quotes for aesthetic reasons, similar to extra whitespace. And the whitespace settings are global.

**Comment by pjkaufman** [comment]
I know it has been a while on this. But I finally took a stab at this and should have a fix for it. It will go out with the next release.

**Comment by pjkaufman** [comment]
The change has been merged into master and should go out with the next release. Please let us know if the bug around properly recognizing commas or the lack of removing unnecessary escape chars for multi-line arrays does not work when the setting to do so is turned on for master or the next release.
